File: app/schemas/incident.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

from app.models.enums import EventSeverity, EventStatus


class ZabbixWebhookPayload(BaseModel):
    event_id: str
    trigger_id: Optional[str] = None
    trigger_name: str
    trigger_severity: str  # Zabbix API: "Disaster"|"High"|"Average"|"Warning"|"Information"
    trigger_status: str  # Zabbix API: "PROBLEM"|"RESOLVED"
    host_name: str
    host_ip: Optional[str] = None
    event_date: Optional[str] = None
    event_time: Optional[str] = None
    event_value: Optional[str] = None  # Zabbix API: "1"=problem, "0"=resolved
    item_name: Optional[str] = None
    item_value: Optional[str] = None
    event_tags: Optional[str] = None
    event_message: Optional[str] = None


class WebhookResponse(BaseModel):
    status: str
    occurrence_id: int
    action: str
    message: str


class SeverityCount(BaseModel):
    severity: str
    count: int


class StatusCount(BaseModel):
    status: str
    count: int


class TopRiskEvent(BaseModel):
    id: int
    event_name: str
    severity: str
    host: Optional[str]
    occurrence_count: int
    risk_score: Optional[int]
    current_status: str
    first_seen_at: datetime


class RecentActivity(BaseModel):
    id: int
    event_name: str
    action: str
    actor: Optional[str]
    timestamp: datetime


class DashboardResponse(BaseModel):
    total_events: int
    open_events: int
    critical_events: int
    resolved_today: int
    severity_distribution: List[SeverityCount]
    status_distribution: List[StatusCount]
    top_risk_events: List[TopRiskEvent]
    recent_activities: List[RecentActivity]
    recurring_event_count: int


class EventListItem(BaseModel):
    id: int
    event_name: str
    severity: str
    host: Optional[str]
    service: Optional[str]
    current_status: str
    first_seen_at: datetime
    last_seen_at: Optional[datetime]
    occurrence_count: int
    source_system: str
    has_assessment: bool
    has_related_docs: bool
    risk_score: Optional[int] = None
    recurrence_score: Optional[int] = None

    model_config = {"from_attributes": True}


class EventListResponse(BaseModel):
    events: List[EventListItem]
    total: int
    page: int
    page_size: int


class StateHistoryItem(BaseModel):
    id: int
    previous_state: Optional[str]
    new_state: str
    changed_by: Optional[str]
    source: Optional[str]
    description: Optional[str]
    changed_at: datetime

    model_config = {"from_attributes": True}


class HandlingRecordItem(BaseModel):
    id: int
    action_type: str
    action_summary: str
    action_details: Optional[str]
    actor: str
    executed_at: datetime
    related_ticket: Optional[str]
    result_status: Optional[str]

    model_config = {"from_attributes": True}


class AssessmentDetail(BaseModel):
    id: int
    recurrence_score: Optional[int]
    risk_score: Optional[int]
    pattern_summary: Optional[str]
    probable_cause: Optional[str]
    transfer_to_incident: bool
    analyzed_at: Optional[datetime]
    analyzer_type: Optional[str]

    model_config = {"from_attributes": True}


class RelatedDocumentBrief(BaseModel):
    id: int
    title: str
    document_type: Optional[str]
    protection_type: str
    content_available: bool
    refined_available: bool
    limitation: Optional[str] = None


class RelatedIncidentBrief(BaseModel):
    id: int
    title: str
    severity: Optional[str]
    occurred_at: Optional[datetime]
    resolution_summary: Optional[str]


class MetricLogEvidenceItem(BaseModel):
    id: int
    evidence_type: str
    source: str
    summary: str
    collected_at: datetime

    model_config = {"from_attributes": True}


class EventDetailResponse(BaseModel):
    id: int
    event_name: str
    severity: str
    host: Optional[str]
    service: Optional[str]
    current_status: str
    source_system: str
    source_event_id: str
    first_seen_at: datetime
    last_seen_at: Optional[datetime]
    occurrence_count: int
    raw_payload_reference: Optional[str]
    created_at: datetime
    state_history: List[StateHistoryItem]
    handling_records: List[HandlingRecordItem]
    assessment: Optional[AssessmentDetail]
    metric_log_evidence: List[MetricLogEvidenceItem]
    related_documents: List[RelatedDocumentBrief]
    related_incidents: List[RelatedIncidentBrief]
    recommended_actions: List[str]
    limitation_flags: List[str]


class AnalysisReportResponse(BaseModel):
    occurrence_id: int
    event_name: str
    generated_at: datetime
    analyzer_type: str
    summary: str
    probable_causes: List[str]
    risk_level: str
    risk_score: int = Field(ge=0, le=100)
    recurrence_analysis: str
    recurrence_score: int = Field(ge=0, le=100)
    recommended_actions: List[str]
    related_patterns: List[str]
    evidence_used: List[dict]
    limitation_flags: List[str]


class EventChatMessage(BaseModel):
    role: str
    content: str


class EventChatRequest(BaseModel):
    query: str
    conversation_history: Optional[List[EventChatMessage]] = None
    use_sanitized_knowledge: bool = True


class EventChatEvidence(BaseModel):
    source_id: str
    source_type: str
    title: str
    content_preview: Optional[str] = None
    relevance_score: float = Field(ge=0.0, le=1.0, default=0.5)


class EventChatResponse(BaseModel):
    answer: str
    evidence: List[EventChatEvidence]
    limitation_notice: Optional[str] = None
    limitation_flags: List[str] = []
    requires_manual_confirmation: bool = False
    conversation_id: Optional[str] = None

File: frontend/src/types/incident.ts
export type EventSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info'
export type EventStatus = 'open' | 'acknowledged' | 'resolved' | 'closed' | 'reopened'

export interface SeverityCount {
  severity: string
  count: number
}

export interface StatusCount {
  status: string
  count: number
}

export interface TopRiskEvent {
  id: number
  event_name: string
  severity: string
  host: string | null
  occurrence_count: number
  risk_score: number | null
  current_status: string
  first_seen_at: string
}

export interface RecentActivity {
  id: number
  event_name: string
  action: string
  actor: string | null
  timestamp: string
}

export interface DashboardData {
  total_events: number
  open_events: number
  critical_events: number
  resolved_today: number
  severity_distribution: SeverityCount[]
  status_distribution: StatusCount[]
  top_risk_events: TopRiskEvent[]
  recent_activities: RecentActivity[]
  recurring_event_count: number
}

export interface EventListItem {
  id: number
  event_name: string
  severity: string
  host: string | null
  service: string | null
  current_status: string
  first_seen_at: string
  last_seen_at: string | null
  occurrence_count: number
  source_system: string
  has_assessment: boolean
  has_related_docs: boolean
  risk_score: number | null
  recurrence_score: number | null
}

export interface EventListResponse {
  events: EventListItem[]
  total: number
  page: number
  page_size: number
}

export interface StateHistoryItem {
  id: number
  previous_state: string | null
  new_state: string
  changed_by: string | null
  source: string | null
  description: string | null
  changed_at: string
}

export interface HandlingRecordItem {
  id: number
  action_type: string
  action_summary: string
  action_details: string | null
  actor: string
  executed_at: string
  related_ticket: string | null
  result_status: string | null
}

export interface AssessmentDetail {
  id: number
  recurrence_score: number | null
  risk_score: number | null
  pattern_summary: string | null
  probable_cause: string | null
  transfer_to_incident: boolean
  analyzed_at: string | null
  analyzer_type: string | null
}

export interface RelatedDocumentBrief {
  id: number
  title: string
  document_type: string | null
  protection_type: string
  content_available: boolean
  refined_available: boolean
  limitation: string | null
}

export interface RelatedIncidentBrief {
  id: number
  title: string
  severity: string | null
  occurred_at: string | null
  resolution_summary: string | null
}

export interface MetricLogEvidenceItem {
  id: number
  evidence_type: string
  source: string
  summary: string
  collected_at: string
}

export interface EventDetail {
  id: number
  event_name: string
  severity: string
  host: string | null
  service: string | null
  current_status: string
  source_system: string
  source_event_id: string
  first_seen_at: string
  last_seen_at: string | null
  occurrence_count: number
  raw_payload_reference: string | null
  created_at: string
  state_history: StateHistoryItem[]
  handling_records: HandlingRecordItem[]
  assessment: AssessmentDetail | null
  metric_log_evidence: MetricLogEvidenceItem[]
  related_documents: RelatedDocumentBrief[]
  related_incidents: RelatedIncidentBrief[]
  recommended_actions: string[]
  limitation_flags: string[]
}

export interface AnalysisReport {
  occurrence_id: number
  event_name: string
  generated_at: string
  analyzer_type: string
  summary: string
  probable_causes: string[]
  risk_level: string
  risk_score: number
  recurrence_analysis: string
  recurrence_score: number
  recommended_actions: string[]
  related_patterns: string[]
  evidence_used: Record<string, unknown>[]
  limitation_flags: string[]
}

export interface ChatEvidence {
  source_id: string
  source_type: string
  title: string
  content_preview: string | null
  relevance_score: number
}

export interface ChatResponse {
  answer: string
  evidence: ChatEvidence[]
  limitation_notice: string | null
  limitation_flags: string[]
  requires_manual_confirmation: boolean
  conversation_id: string | null
}

export interface ChatMessage {
  role: 'user' | 'assistant'
  content: string
}
